nixos/modules/services/hardware/undervolt.nix

```diff
@@ -0,0 +1,92 @@
+{ lib, pkgs, config, ... }:
+
+with lib;
+let
+  cfg = config.services.undervolt;
+in {
+  options = {
+    services = {
+      undervolt = {
+        enable = mkOption {
+          type = types.bool;
+          default = false;
+          description = ''
+            Whether to enable the undervolt daemon.
+          '';
+        };
+
+        core = mkOption {
+          type = types.int;
+          default = 0;
+          description = ''
+            Core offset (mV)
+          '';
+        };
+
+        cache = mkOption {
+          type = types.int;
+          default = 0;
+          description = ''
+            Cache offset (mV)
+          '';
+        };
+
+        analogio = mkOption {
+          type = types.int;
+          default = 0;
+          description = ''
+            Analogio offset (mV)
+          '';
+        };
+
+        uncore = mkOption {
+          type = types.int;
+          default = 0;
+          description = ''
+            Uncore offset (mV)
+          '';
+        };
+
+        gpu = mkOption {
+          type = types.int;
+          default = 0;
+          description = ''
+            Gpu offset (mV)
+          '';
+        };
+      };
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    environment = {
+      systemPackages = with pkgs; [
+        undervolt
+      ];
+    };
+
+    systemd = {
+      services = {
+        undervolt = {
+          description = "Intel undervolting";
+          after = [
+            "suspend.target"
+            "hibernate.target"
+            "hybrid-sleep.target"
+          ];
+          serviceConfig = {
+            ExecStart = ''
+              ${pkgs.undervolt}/bin/undervolt --core ${toString cfg.core} --cache ${toString cfg.cache} --analogio ${toString cfg.analogio} --uncore ${toString cfg.uncore} --gpu ${toString cfg.gpu}
+            '';
+          };
+          wantedBy = [
+            "suspend.target"
+            "hibernate.target"
+            "hybrid-sleep.target"
```

Comment:
Did you test this? Aren't those three targets activated before the actual suspend/sleep?

The `powerManagement.resumeCommands` option is probably more useful.

Reply:
Moved cfg under `powerManagement` while also using `powerManagement.powerUpCommands`.

Reply:
Nevermind, `powerManagement.powerUpCommands` fails when booting up while `powerManagement.resumeCommands` doesn't run at boot. Just returned back to the service.